### PR TITLE
Convert firrtl-interpreter to use new options as annotations.

### PR DESCRIPTION
**Pull request by chick**
Massive changes along the lines of what was done for treadle.
See Firrtl RFC Issue #764 Annotations subsume ExecutionOptions?
MERGE only after merging [Firrtl PR #765](https://github.com/freechipsproject/firrtl/pull/765)

### CODE REVIEW

File: src/main/scala/firrtl_interpreter/InterpreterOptions.scala [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+import firrtl.AnnotationSeq
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.options.{HasScoptOptions, RegisteredLibrary}
+import scopt.OptionParser
+
+sealed trait InterpreterOption extends HasScoptOptions
+
+/**
+  * Tells treadle to write a vcd file during simulation
```
Comment (ucbjrl) [inline src/main/scala/firrtl_interpreter/InterpreterOptions.scala]: Cut 'n paste - treadle -> firrtl interpreter

File: src/main/scala/firrtl_interpreter/InterpretiveTester.scala [hunk 2]
```diff
@@ -31,13 +49,22 @@ class InterpretiveTester(input: String, optionsManager: HasInterpreterSuite = ne
...
+
+    interpreter.makeVCDLogger(updatedOptions.getBuildFileName("vcd"), interpreterOptions.vcdShowUnderscored)
   }
+//    optionsManager.setTopNameIfNotSet(interpreter.loweredAst.main)
+//    optionsManager.makeTargetDir()
+//    interpreter.makeVCDLogger(
+//      interpreterOptions.vcdOutputFileName(optionsManager),
+//      interpreterOptions.vcdShowUnderscored
+//    )
+//  }
```
Comment (ucbjrl) [inline src/main/scala/firrtl_interpreter/InterpretiveTester.scala]: Remove commented-out code?

File: src/main/scala/firrtl_interpreter/Driver.scala [hunk 3]
```diff
@@ -2,124 +2,44 @@
...
-    }
-    .text("depth of stack used to evaluate expressions")
-
-  parser.opt[Unit]("show-firrtl-at-load")
-    .abbr("fisfas")
-    .foreach { _ =>
-      interpreterOptions = interpreterOptions.copy(showFirrtlAtLoad = true)
-    }
-    .text("compiled low firrtl at firrtl load time")
+object Driver {
```
Comment (seldridge) [inline src/main/scala/firrtl_interpreter/Driver.scala]: Any motivation for not having this extend `firrtl.options.Driver`?

File: src/test/scala/firrtl_interpreter/DriverSpec.scala [hunk 4]
```diff
@@ -15,15 +16,15 @@ class DriverSpec extends FreeSpec with Matchers {
           |    output y : UInt<1>
           |    y <= x
         """.stripMargin
-      //      val interpreter = Driver.execute(Array.empty[String], input)
-      val interpreter = Driver.execute(Array("--fint-verbose"), input)
 
-      interpreter should not be empty
+      val source = FirrtlSourceAnnotation(input)
+      val result = Driver.execute(Array("--fint-verbose"), Seq(source))
```
Comment (ucbjrl) [inline src/test/scala/firrtl_interpreter/DriverSpec.scala]: This appears to be causing:
```bash
Error: Unknown option --fint-verbose
Try --help for more information.

Failed to parse command line options
firrtl.FIRRTLException: Failed to parse command line options
	at firrtl.options.ExecutionOptionsManager$$anonfun$parse$1.apply(ExecutionOptionsManager.scala:69)
	at firrtl.options.ExecutionOptionsManager$$anonfun$parse$1.apply(ExecutionOptionsManager.scala:69)
	at scala.Option.getOrElse(Option.scala:121)
	at firrtl.options.ExecutionOptionsManager.parse(ExecutionOptionsManager.scala:69)
	at firrtl_interpreter.Driver$.execute(Driver.scala:33)
	at firrtl_interpreter.DriverSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(DriverSpec.scala:21)
	at firrtl_interpreter.DriverSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(DriverSpec.scala:10)
	at firrtl_interpreter.DriverSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(DriverSpec.scala:10)
 ...
```

Comment (seldridge) [reply src/test/scala/firrtl_interpreter/DriverSpec.scala]: This PR appears to be missing the `META-INF` `ServiceLoader` registration... PR inbound. (There's a short option duplication after that that I'm looking into.)

Comment (seldridge) [reply src/test/scala/firrtl_interpreter/DriverSpec.scala]: This referenced "short option duplication" is actually long option duplication. HEAD of https://github.com/freechipsproject/firrtl/pull/765 has an up to date version that shows the correct error.

File: src/main/scala/firrtl_interpreter/vcd/VCDConfig.scala [hunk 5]
```diff
@@ -2,54 +2,102 @@
...
 
-  parser.note("vcd")
+sealed trait VcdConfigOption extends HasScoptOptions
 
-  parser.opt[String]("start-scope")
-    .abbr("vss")
-    .foreach { x => vcdConfig = vcdConfig.copy(startScope = x)}
-    .text("starts saving information at specified scope")
+case class VcdSourceNameAnnotation(name: String = "") extends NoTargetAnnotation with VcdConfigOption {
+  def addOptions(p: OptionParser[AnnotationSeq]): Unit = p.opt[String]("vcd-start-scope")
```
Comment (seldridge) [inline src/main/scala/firrtl_interpreter/vcd/VCDConfig.scala]: This option is a duplicate of the long option for `VcdStartScopeAnnotation`.

File: src/main/scala/firrtl_interpreter/VcdReplayOptions.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    .abbr("vr-evtr")
+    .action( (events, c) => c :+ VcdReplayEventsToRunAnnotation(events) )
+    .unbounded()
+
+    .text("events to run")
+}
+
+case object VcdReplayTestAliasedWiresAnnotation extends NoTargetAnnotation with VcdReplayOption {
+  def addOptions(p: OptionParser[AnnotationSeq]): Unit = p.opt[Unit]("vcd-replay-test-aliased-wires")
+    .abbr("vr-evtr")
```
Comment (seldridge) [inline src/main/scala/firrtl_interpreter/VcdReplayOptions.scala]: This is a duplicate of above.

File: src/main/scala/firrtl_interpreter/VcdReplayOptions.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    .abbr("vr-vf")
+    .action( (name, c) => c :+ VcdReplayVcdFileAnnotation(name) )
+    .unbounded()
+
+    .text("vcd source for vcd replay to operate on")
+}
+
+case class VcdReplaySkipEventsAnnotation(skip: Int = 0) extends NoTargetAnnotation with VcdReplayOption {
+  def addOptions(p: OptionParser[AnnotationSeq]): Unit = p.opt[Int]("vcd-replay-skip-events")
+    .abbr("vr-vf")
```
Comment (seldridge) [inline src/main/scala/firrtl_interpreter/VcdReplayOptions.scala]: Duplicate of above.

File: src/main/scala/firrtl_interpreter/repl/ReplConfig.scala [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  def addOptions(p: OptionParser[AnnotationSeq]): Unit = p.opt[String]("irepl-script-name")
+    .abbr("ir-sf")
+    .action( (name, c) => c :+ ReplScriptNameAnnotation(name) )
+    .unbounded()
+    .text("script file to load")
+}
+
+case object ReplUseVcdScriptAnnotation extends NoTargetAnnotation with ReplOption {
+  def addOptions(p: OptionParser[AnnotationSeq]): Unit = p.opt[Unit]("irepl-use-vcd")
+    .abbr("ir-sf")
```
Comment (seldridge) [inline src/main/scala/firrtl_interpreter/repl/ReplConfig.scala]: This is duplicated three times in total. Note: totally up to you if you even want to have short options. It gets complicated to manage them all...

File: src/main/scala/firrtl_interpreter/InterpreterOptions.scala (hunk 1 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/InterpreterOptions.scala]: Good catch, done

File: src/main/scala/firrtl_interpreter/InterpretiveTester.scala (hunk 2 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/InterpretiveTester.scala]: done

File: src/main/scala/firrtl_interpreter/Driver.scala (hunk 3 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/Driver.scala]: Nope, just a miss.

File: src/main/scala/firrtl_interpreter/vcd/VCDConfig.scala (hunk 5 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/vcd/VCDConfig.scala]: fixed typo in VcdSourceNameAnnotation

File: src/main/scala/firrtl_interpreter/repl/ReplConfig.scala (hunk 8 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/repl/ReplConfig.scala]: just sloppy. I have dusted off my old `grep ... | cut -d: -f1 | sort | uniq -c | sort -n` command and think I have fixed them all now

File: src/main/scala/firrtl_interpreter/VcdReplayOptions.scala (hunk 7 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/VcdReplayOptions.scala]: done

File: src/main/scala/firrtl_interpreter/VcdReplayOptions.scala (hunk 6 above)
Comment (chick) [reply src/main/scala/firrtl_interpreter/VcdReplayOptions.scala]: done

File: src/test/scala/firrtl_interpreter/DriverSpec.scala (hunk 4 above)
Comment (chick) [reply src/test/scala/firrtl_interpreter/DriverSpec.scala]: merged @seldridge PR that fixes this. Thanks!

**Comment by chick** [comment]
This will be re-done using new Stage/Shell implementation

**Pull request closed without merging**